### PR TITLE
Add bit_xor Spark aggregate function

### LINKED ISSUE #4462: Add bit_xor function from Spark
### Description

Spark support bit_xor aggregate function since 3.0.0 [docs](https://spark.apache.org/docs/3.3.1/api/sql/#bit_xor), I would implement this function in Velox.

Please assign to me, thanks!

### PR DESCRIPTION
**Pull request by Yohahaha**
Add bit_xor Spark aggregate function: https://spark.apache.org/docs/3.3.1/api/sql/#bit_xor.

Extract common bitwise aggregate logic into lib/BitwiseAggregateBase.h

Add SparkAggregationFuzzerTest.cpp to run fuzzer test.

Fixes #4462

### CODE REVIEW

File: velox/docs/functions/spark/aggregate.rst [hunk 1]
```diff
@@ -15,3 +15,7 @@ General Aggregate Functions
 
     Returns the last value of `x` for a group of rows. If `ignoreNull` is true, returns only non-null values.
 
+.. spark:function:: bitwise_xor_agg(x) -> bigint
```
Comment (mbasmanova) [inline velox/docs/functions/spark/aggregate.rst:18]: Please, move this function to preserve alphabetical order.

File: velox/docs/functions/spark/aggregate.rst [hunk 2]
```diff
@@ -15,3 +15,7 @@ General Aggregate Functions
...
+
+    Alias of Spark `bit_xor` function.
```
Comment (mbasmanova) [inline velox/docs/functions/spark/aggregate.rst:20]: Why not name this function "bit_xor"?

File: velox/functions/prestosql/aggregates/BitwiseAggregateBase.h [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+#include "velox/exec/Aggregate.h"
+#include "velox/expression/FunctionSignature.h"
+#include "velox/functions/prestosql/aggregates/AggregateNames.h"
+#include "velox/functions/prestosql/aggregates/SimpleNumericAggregate.h"
+
+namespace facebook::velox::aggregate::prestosql {
+
+template <typename T>
+class BitwiseAndOrXorAggregate : public SimpleNumericAggregate<T, T, T> {
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/BitwiseAggregateBase.h:26]: perhaps, BitwiseAggregateBase

File: velox/functions/prestosql/aggregates/BitwiseAggregateBase.h [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    this->addSingleGroupRawInput(group, rows, args, mayPushdown);
+  }
+
+ protected:
+  const T initialValue_;
+};
+
+template <template <typename U> class T>
+bool registerBitwise(const std::string& name) {
+  // TODO Fix the signatures to match Presto.
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/BitwiseAggregateBase.h:75]: This TODO is no longer needed. Let's remove.

File: velox/functions/prestosql/aggregates/CMakeLists.txt [hunk 5]
```diff
@@ -21,6 +21,7 @@ add_library(
   ArbitraryAggregate.cpp
   ArrayAggAggregate.cpp
   AverageAggregate.cpp
+  BitwiseAggregateBase.h
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/CMakeLists.txt:24]: Do not add .h files to CMakeLists.txt

File: velox/functions/sparksql/aggregates/BitwiseXorAggregate.cpp [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        [](T& result, T value) { result ^= value; },
+        [](T& result, T value, int /* unused */
+        ) { result ^= value; },
+        mayPushdown,
+        this->initialValue_);
+  }
+};
+
+template <template <typename U> class T>
+bool registerBitwise(const std::string& name) {
```
Comment (mbasmanova) [inline velox/functions/sparksql/aggregates/BitwiseXorAggregate.cpp:65]: Should we reuse this function from BitwiseAggregateBase.h?

**Review by mbasmanova — COMMENTED** [review]
@Yohahaha Thank you for adding this aggregate function. Looks good overall. Some comments below.

Do we have Fuzzer for Spark aggregates? If not, let's setup one. See velox/expression/tests/SparkExpressionFuzzerTest.cpp and velox/exec/tests/AggregationFuzzerTest.cpp

File: velox/functions/sparksql/aggregates/BitwiseXorAggregate.cpp [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/sparksql/aggregates/BitwiseXorAggregate.h"
+
+#include "velox/functions/prestosql/aggregates/BitwiseAggregateBase.h"
```
[the remark this answers is not in the document]
Comment (Yohahaha) [reply velox/functions/sparksql/aggregates/BitwiseXorAggregate.cpp]: > Would velox/functions/lib/aggregates work?

I will try to refactor it.

File: velox/exec/tests/SparkAggregationFuzzerTest.cpp [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+      "last" // Not support TIMESTAMP.
+  };
+
+  // The results of the following functions depend on the order of input
+  // rows. For some functions, the result can be transformed to a value that
+  // doesn't depend on the order of inputs. If such transformation exists, it
+  // can be specified to be used for results verification. If no transformation
+  // is specified, results are not verified.
+  std::unordered_map<std::string, std::string> orderDependentFunctions = {
+      {"bit_xor", ""}
```
Comment (mbasmanova) [inline velox/exec/tests/SparkAggregationFuzzerTest.cpp]: Why is bit_xor order-dependent?

File: velox/functions/lib/BitwiseAggregateBase.h [hunk 9]
```diff
@@ -15,19 +15,18 @@
  */
 #pragma once
 
+#include "SimpleNumericAggregate.h"
```
Comment (mbasmanova) [inline velox/functions/lib/BitwiseAggregateBase.h:18]: Use full paths.

File: velox/functions/prestosql/aggregates/ArbitraryAggregate.cpp [hunk 10]
```diff
@@ -16,8 +16,8 @@
 
 #include "velox/exec/Aggregate.h"
 #include "velox/expression/FunctionSignature.h"
+#include "velox/functions/lib/SimpleNumericAggregate.h"
 #include "velox/functions/prestosql/aggregates/AggregateNames.h"
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/ArbitraryAggregate.cpp:20]: Do we need this include? Can we remove it?

File: velox/functions/prestosql/aggregates/ArbitraryAggregate.cpp [hunk 11]
```diff
@@ -16,8 +16,8 @@
...
-#include "velox/functions/prestosql/aggregates/SimpleNumericAggregate.h"
 #include "velox/functions/prestosql/aggregates/SingleValueAccumulator.h"
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/ArbitraryAggregate.cpp:21]: Should we move SingleValueAccumulator as well? We don't want to have "lib" depend on "prestosql".

File: velox/exec/tests/SparkAggregationFuzzerTest.cpp [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+  // Calls common init functions in the necessary order, initializing
+  // singletons, installing proper signal handlers for better debugging
+  // experience, and initialize glog and gflags.
+  folly::init(&argc, &argv);
+
+  // TODO: List of the functions that at some point crash or fail and need to
+  // be fixed before we can enable.
+  std::unordered_set<std::string> skipFunctions = {
+      "last" // Not support TIMESTAMP.
```
Comment (mbasmanova) [inline velox/exec/tests/SparkAggregationFuzzerTest.cpp:52]: Maybe create GitHub issue with fuzzer-found label and add a link to it here? Would you be willing to help fix this problem?

Comment (Yohahaha) [reply velox/exec/tests/SparkAggregationFuzzerTest.cpp:52]: Sure, I will create issue soon with fuzzer test error log, and try to fix it.
Let's skip it for now.

File: velox/exec/tests/SparkAggregationFuzzerTest.cpp (hunk 8 above)
Comment (Yohahaha) [reply velox/exec/tests/SparkAggregationFuzzerTest.cpp]: I got it, will be fixed.

File: velox/functions/prestosql/aggregates/ArbitraryAggregate.cpp (hunk 10 above)
Comment (Yohahaha) [reply velox/functions/prestosql/aggregates/ArbitraryAggregate.cpp:20]: I think we cannot remove it, it define functions name.

**Comment by Yohahaha** [comment]
Hi @mbasmanova 
I move bitwise shared code into `functions/lib` for now.

If we want to extract shared aggregate code into `functions/lib/aggregates`, we have to create a new static library like `velox_functions_aggregate_common`, and rename current static library name under `functions/prestosql` such as `velox_aggregates` to `velox_functions_aggregates` or `velox_functions_prestosql_aggregates` to make library naming more reasonable.

It would better open another PR to do this code refactor if you like.

File: velox/exec/tests/SparkAggregationFuzzerTest.cpp (hunk 12 above)
Comment (Yohahaha) [reply velox/exec/tests/SparkAggregationFuzzerTest.cpp:52]: See #4482 .

**Review by mbasmanova — APPROVED** [review]
@Yohahaha Looks good to me but CI is failing. Please, take a look.

@kgpai Krishna, would you help review CI config changes?

**Comment by Yohahaha** [comment]
Hi @mbasmanova 
> Looks good to me but CI is failing. Please, take a look.

CI failed due to network problem, I will trigger again when network be good.

Let me run a 1 hour fuzzer test.

**Review by kgpai — APPROVED** [review]

**Comment by kgpai** [comment]
@Yohahaha Import latest main to fix the builds - arrow urls have changed.

**Comment by Yohahaha** [comment]
This 1 hour fuzzer test is OK.
`./velox/exec/tests/spark_aggregation_fuzzer_test --seed 123 --duration_sec 3600 --logtostderr=1 --minloglevel=0 --repro_persist_path=/tmp/aggregate_fuzzer_repro --num_batches=1`
```
I0404 00:20:49.380654 1456174 AggregationFuzzer.cpp:604] ==============================> Done with iteration 42726
I0404 00:20:49.380683 1456174 AggregationFuzzer.cpp:1049] Total functions tested: 1
I0404 00:20:49.380685 1456174 AggregationFuzzer.cpp:1050] Total masked aggregations: 7615 (17.82%)
I0404 00:20:49.380694 1456174 AggregationFuzzer.cpp:1052] Total global aggregations: 3851 (9.01%)
I0404 00:20:49.380698 1456174 AggregationFuzzer.cpp:1054] Total group-by aggregations: 34650 (81.10%)
I0404 00:20:49.380702 1456174 AggregationFuzzer.cpp:1056] Total distinct aggregations: 0 (0.00%)
I0404 00:20:49.380707 1456174 AggregationFuzzer.cpp:1058] Total window expressions: 4226 (9.89%)
I0404 00:20:49.380709 1456174 AggregationFuzzer.cpp:1060] Total aggregations verified against DuckDB: 30482 (71.34%)
I0404 00:20:49.380713 1456174 AggregationFuzzer.cpp:1062] Total failed aggregations: 0 (0.00%)
```

But I got a error of Window.
`./velox/exec/tests/spark_aggregation_fuzzer_test --seed 1848296956 --duration_sec 10 --logtostderr=1 --minloglevel=0 --repro_persist_path=/tmp/aggregate_fuzzer_repro`
```
I0404 09:55:42.165649 3411656 AggregationFuzzer.cpp:251] Total functions: 1 (4 signatures)
I0404 09:55:42.165830 3411656 AggregationFuzzer.cpp:253] Functions with at least one supported signature: 1 (100.00%)
I0404 09:55:42.165841 3411656 AggregationFuzzer.cpp:257] Functions with no supported signature: 0 (0.00%)
I0404 09:55:42.165845 3411656 AggregationFuzzer.cpp:259] Supported function signatures: 4 (100.00%)
I0404 09:55:42.165848 3411656 AggregationFuzzer.cpp:263] Unsupported function signatures: 0 (0.00%)
I0404 09:55:42.165864 3411656 AggregationFuzzer.cpp:532] ==============================> Started iteration 0 (seed: 1848296956)
I0404 09:55:42.175367 3411656 AggregationFuzzer.cpp:624] Executing query plan:
-- Window[partition by [p0] order by [s0 ASC NULLS LAST, s1 ASC NULLS LAST, s2 ASC NULLS LAST, s3 ASC NULLS LAST, s4 ASC NULLS LAST] w0 := bit_xor(ROW["c0"]) RANGE between UNBOUNDED PRECEDING and CURRENT ROW] ->
c0:INTEGER, p0:BOOLEAN, s0:VARBINARY, s1:DATE, s2:TINYINT, s3:INTEGER, s4:VARCHAR, row_number:BIGINT, w0:INTEGER
  -- Values[1000 rows in 10 vectors] -> c0:INTEGER, p0:BOOLEAN, s0:VARBINARY, s1:DATE, s2:TINYINT, s3:INTEGER, s4:VARCHAR, row_number:BIGINT
I0404 09:55:42.192199 3411672 Task.cpp:825] All drivers (1) finished for task test_cursor 1 after running for 17 ms.
I0404 09:55:42.192220 3411672 Task.cpp:1474] Terminating task test_cursor 1 with state Finished after running for 17 ms.
I0404 09:55:42.194061 3411656 AggregationFuzzer.cpp:639] [ROW ROW<c0:INTEGER,p0:BOOLEAN,s0:VARBINARY,s1:DATE,s2:TINYINT,s3:INTEGER,s4:VARCHAR,row_number:BIGINT,w0:INTEGER>: 1000 elements, no nulls]
I0404 09:55:42.194113 3411656 AggregationFuzzer.cpp:893] SELECT c0, p0, s0, s1, s2, s3, s4, row_number, bit_xor(c0) OVER (partition by p0 order by s0, s1, s2, s3, s4) FROM tmp
../velox/exec/tests/utils/QueryAssertions.cpp:1076: Failure
Failed
Expected 1000, got 1000
1 extra rows, 1 missing rows
1 of extra rows:
        1100464885 | false | "thK4&Wn:Alavd)^7'sLNL(L5v7*\\DnfcJSZ$~H&_q.bLD!PuAL+>!|PLiSryw8$" | "909831-02-23" | 4 | 638228802 | null | 570 | 458356396

1 of missing rows:
        1100464885 | false | "thK4&Wn:Alavd)^7'sLNL(L5v7*\\DnfcJSZ$~H&_q.bLD!PuAL+>!|PLiSryw8$" | "909831-02-23" | 4 | 638228802 | null | 570 | 1461150393

Unexpected results
E0404 09:55:42.450125 3411656 Exceptions.h:68] Line: ../velox/exec/tests/AggregationFuzzer.cpp:961, Function:verifyWindow, Expression: assertEqualResults(expectedResult.value(), {resultOrError.result}) Velox and
DuckDB results don't match, Source: RUNTIME, ErrorCode: INVALID_STATE
I0404 09:55:42.451520 3411656 AggregationFuzzer.cpp:452] Persisted input: /tmp/aggregate_fuzzer_repro/velox_vector_ojhYVB and plan: /tmp/aggregate_fuzzer_repro/velox_plan_9RJKWC
terminate called after throwing an instance of 'facebook::velox::VeloxRuntimeError'
  what():  Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Velox and DuckDB results don't match
Retriable: False
Expression: assertEqualResults(expectedResult.value(), {resultOrError.result})
Function: verifyWindow
File: ../velox/exec/tests/AggregationFuzzer.cpp
Line: 961
```


I would find the reason and try to fix it.
cc @mbasmanova

**Comment by Yohahaha** [comment]
> But I got a error of Window.
./velox/exec/tests/spark_aggregation_fuzzer_test --seed 1848296956 --duration_sec 10 --logtostderr=1 --minloglevel=0 --repro_persist_path=/tmp/aggregate_fuzzer_repro

As #4502 discussed, aggregate function's result for peer rows are same, I also verified this behavior in DuckDB.

But the error of above fuzzer test is DuckDB in Velox for peer rows return different results.

DuckDB results:
```
2035041034 false "rD~FT-{@{oAoaXBj`wk1[mw,fI0YnHwU<VGU,9d0O`Zx" null 47 755690227 "Nq~w3lCr}69\"K%t&nYm|6i4?zMILjmCZZqS9z`,G`GGd~>`^JFm`]AjY;!S/!QDk" 872 377528396
1100464885 false "thK4&Wn:Alavd)^7'sLNL(L5v7*\\DnfcJSZ$~H&_q.bLD!PuAL+>!|PLiSryw8$" "909831-02-23" 4 638228802 null 570 1461150393
1279693845 false "thK4&Wn:Alavd)^7'sLNL(L5v7*\\DnfcJSZ$~H&_q.bLD!PuAL+>!|PLiSryw8$" "909831-02-23" 4 638228802 null 585 458356396
```

Velox results:
```
467: {2035041034, false, rD~FT-{@{oAoaXBj`wk1[mw,fI0YnHwU<VGU,9d0O`Zx, null, 47, 755690227, Nq~w3lCr}69"K%t&nYm|6i4?zMILjmCZZqS9z`,G`GGd~>`^JFm`]AjY;!S/!QDk, 872, 377528396}
468: {1100464885, false, thK4&Wn:Alavd)^7'sLNL(L5v7*\DnfcJSZ$~H&_q.bLD!PuAL+>!|PLiSryw8$, 909831-02-23, 4, 638228802, null, 570, 458356396}
469: {1279693845, false, thK4&Wn:Alavd)^7'sLNL(L5v7*\DnfcJSZ$~H&_q.bLD!PuAL+>!|PLiSryw8$, 909831-02-23, 4, 638228802, null, 585, 458356396}
```

Penultimate value of row is pre-generated row number, let me use illustrate the correct computation procedure:

1. row 570 and 585 have same order by keys, so they are peer rows.
2. compute aggregate functions for peer rows: 1100464885(row 570) bit_xor 1279693845(row 585) = 231823072
3. compute peer rows' agg result with previous non-peer row once: 377528396(row 872) bit_xor 231823072 = 458356396
4. apply above result to each peer row


I'm not sure how DuckDB in Velox works, @mbasmanova could you help take a look on this?

**Comment by mbasmanova** [comment]
@Yohahaha 

It looks like a bug in the fuzzer.

Velox plan specifies "RANGE between UNBOUNDED PRECEDING and CURRENT ROW" frame:

> -- Window[partition by [p0] order by [s0 ASC NULLS LAST, s1 ASC NULLS LAST, s2 ASC NULLS LAST, s3 ASC NULLS LAST, s4 ASC NULLS LAST] w0 := bit_xor(ROW["c0"]) RANGE between UNBOUNDED PRECEDING and CURRENT ROW] ->

While DuckDB SQL doesn't specify the frame. In Presto, "RANGE between UNBOUNDED PRECEDING and CURRENT ROW" is the default frame, but it could be that in DuckDB the default is "ROWS ...". I looked at DuckDB docs, but could't find any mention of what the default frame is. 

> SELECT c0, p0, s0, s1, s2, s3, s4, row_number, bit_xor(c0) OVER (partition by p0 order by s0, s1, s2, s3, s4) FROM tmp

- https://prestodb.io/docs/current/functions/window.html
- https://duckdb.org/docs/sql/window_functions.html

I suggest to modify the fuzzer to add explicit frame clause to DuckDB SQL. This would be somewhere in makeDuckWindowSql in velox/exec/tests/AggregationFuzzer.cpp

CC: @aditi-pandit

**Comment by mbasmanova** [comment]
> Hi @mbasmanova
> 
> DuckDB window default frame is RANGE, I test it manually.
> 
> I still get same error when specify DuckDB SQL as `SELECT c0, p0, s0, s1, s2, s3, s4, row_number, bit_xor(c0) OVER (partition by p0 order by s0, s1, s2, s3, s4 range between UNBOUNDED PRECEDING and current row) FROM tmp`

Maybe it is a bug in DuckDB. Consider creating an issue in https://github.com/duckdb/duckdb

**Comment by Yohahaha** [comment]
> Maybe it is a bug in DuckDB. Consider creating an issue in https://github.com/duckdb/duckdb

I got it.

I think this PR is complete.

**Review by mbasmanova — APPROVED** [review]

**Comment by mbasmanova** [comment]
@Yohahaha What should we do about the fuzzer failure? Would you like to follow-up with DuckDB folks to confirm whether there is a bug in DuckDB and perhaps get a fix?

**Comment by Yohahaha** [comment]
> Would you like to follow-up with DuckDB folks to confirm whether there is a bug in DuckDB and perhaps get a fix?

Sure, I would try to reproduce in DuckDB side and follow it.

> What should we do about the fuzzer failure?

Could we separate window fuzzer from aggregate fuzzer? add a WindowFuzzerTest?

**Comment by mbasmanova** [comment]
> Could we separate window fuzzer from aggregate fuzzer? add a WindowFuzzerTest?

How would that help?

**Comment by Yohahaha** [comment]
Hi @mbasmanova 

I see CI failed in Facebook Internal Linter, how can I find error log?

**Comment by kgpai** [comment]
@Yohahaha Its not possible to share those signals at this time - fyi in this case though these failures are internal to meta.

**Comment by Yohahaha** [comment]
> @Yohahaha Its not possible to share those signals at this time - fyi in this case though these failures are internal to meta. 

Dose this CI failure blocking merge？ if yes, how can I fix it?

**Comment by Yohahaha** [comment]
CI failed due to `Check function signatures`, but this PR should not cause below problems.
```
Signature removed: [root['to_big_endian_32'], root['from_big_endian_32'], root['to_big_endian_64'], root['from_big_endian_64'], root['from_base64url']]
Found differences: {'dictionary_item_removed': [root['to_big_endian_32'], root['from_big_endian_32'], root['to_big_endian_64'], root['from_big_endian_64'], root['from_base64url']]}
```

@kgpai Could you help to take a look?

**Comment by kgpai** [comment]
Hi @Yohahaha , Just rebase from main and push , the error should go away. I will have a fix out for that soon.

> Dose this CI failure blocking merge？ if yes, how can I fix it?

No someone at Meta will work on that and merge it - no need to worry about that.

**Pull request closed without merging**